ForumSessionProvider/ForumSessionProvider.php

```diff
@@ -18,284 +18,851 @@
+		foreach ($this->FSSettings['SpecialGroups'] as $smf_group_id => $wiki_group_name)
+		{
+			// Group didn't exist?
+			if (!isset($groupActions[$wiki_group_name]))
+				$groupActions[$wiki_group_name] = array();
+
+			// Add the SMF group into the wiki group.
+			$groupActions[$wiki_group_name][] = $smf_group_id;
+		}
+
+		// Now we are going to check all the groups.
+		foreach ($groupActions as $wiki_group_name => $smf_group_ids)
+		{
+			// They are in the SMF group but not the wiki group?
+			if (
+				array_intersect($smf_group_ids, $this->FSgroups) != array()
+				&& in_array($wiki_group_name, $this->userWiki->getEffectiveGroups()) == array()
+			)
+				$this->userWiki->addGroup($wiki_group_name);
+			// They are not in the SMF group, but in the wiki group
+			elseif (
+				array_intersect($smf_group_ids, $this->FSgroups) == array()
+				&& in_array($wiki_group_name, $this->userWiki->getEffectiveGroups()) != array()
+			)
+				$this->userWiki->removeGroup($wiki_group_name);
+		}
+
+		// Did we make any changes?
+		$this->userWiki->setOption('forum_last_update_groups', time());
+		$this->userWiki->saveSettings();
+	}
+
+	/*
+	 * Create a Wiki user
+	*/
+	private function createWikiUser()
+	{
+		$this->userWiki->setName($this->FSuserName);
+		$this->userWiki->setEmail($this->FSuser['email_address']);
+		$this->userWiki->setRealName($this->FSuser['real_name']);
+		$this->userWiki->mEmailAuthenticated = wfTimestampNow();
+
+		$this->userWiki->addToDatabase();
+
+		// Only SMF does this.
+		if ($this->FSSettings['NameStyle'] == 'smf')
+			$user->setOption('forum_member_id', $this->FSuser['id_member']);
+
+		$this->userWiki->setOption('forum_last_update', time());
+		$this->userWiki->saveSettings();
+	}
+
+	// !!! Add checks here for banning
+	private function isBanned()
+	{
+		switch ($this->wgFSPSoftware)
+		{
+			case 'elk1.0':
+			case 'elk1.1':
+				return false;
+				break;
+			case 'smf2.0':
+			case 'smf2.1':
+				return $this->isBannedSMF();
+				break;
+			default: return false;
+		}
+	}
+
+	private function isBannedSMF()
+	{
+		// Check their ban once every 5 minutes.
+		if (!(time() > ((int) $this->userWiki->getOption('forum_last_update_ban', 0) + 300)))
+			return $this->userWiki->getOption('forum_is_banned', 0);
+
+		$request = $this->db_query('
+			SELECT id_ban
+			FROM ' . $this->FSSettings['db_prefix'] . 'ban_items AS i
+			LEFT JOIN ' . $this->FSSettings['db_prefix'] . 'ban_groups AS g
+				ON (i.id_ban_group = g.id_ban_group)
+			WHERE i.id_member = ' . ( (int) $this->FSuser['id_member']) . '
+				AND (g.cannot_post = 1 OR g.cannot_login = 1)');
+
+		$banned = $this->db_num_rows($request);
+		$this->db_free_result($request);
+
+		$this->userWiki->setOption('forum_last_update_ban', time());
+		$this->userWiki->setOption('forum_is_banned', $banned);
+		$this->userWiki->saveSettings();
+
+		return $banned;
+	}
+
+	/**
+	 * Connect to the forum database.
+	 *
+	 * {@source}
+	 */
+	private function FSDBConnect()

```

Comment:
Are there really enough queries to justify incorporating SMF's DB engine, as opposed to just supporting Postgres plain? I fear this will just lead to a mess of trying to support both forums' own DB engines rather than just handling it directly.

Reply:
Either way is fine.  With this syntax we also do get the benefit of using the database layer and its protections.

Reply:
For... two queries, both of which are hunting for an integer, I think we can trust ourselves to cast that.  Best just to load PDO and switch to that where needed.

Reply:
I will try to get time this weekend to change it up to a PDO setup.